# Fama French Regression in Python

#### r - rf = alpha + beta (return_market - rf) + s_iSMB + h_iHML + error term

In [17]:
%pip install yfinance
%pip install statsmodels
# size and value effect

Defaulting to user installation because normal site-packages is not writeable
Looking in links: /usr/share/pip-wheels
Note: you may need to restart the kernel to use updated packages.
Defaulting to user installation because normal site-packages is not writeable
Looking in links: /usr/share/pip-wheels
Note: you may need to restart the kernel to use updated packages.


In [18]:
import yfinance as yf
import datetime as dt
import matplotlib.pyplot as plt
import statsmodels.api as sm

In [19]:
end = dt.date(2025, 6,30)
start = dt.date(end.year - 5, end.month, end.day)

In [20]:
#Fidelity Growth Company Fund - highly diversified (433 holdings)
funds = ['FDGRX']

In [21]:
df = yf.download(funds, start,end, auto_adjust=False)['Adj Close']
df

# We need daily returns 
df = yf.download(funds, start,end, auto_adjust=False)['Adj Close'].pct_change()
fundsret = df
fundsret = fundsret.dropna(axis=0)
fundsret

[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed


Ticker         FDGRX
Date                
2020-07-01  0.013601
2020-07-02  0.004845
2020-07-06  0.019288
2020-07-07 -0.003275
2020-07-08  0.016429
...              ...
2025-06-23  0.001994
2025-06-24  0.016418
2025-06-25  0.005139
2025-06-26  0.010226
2025-06-27  0.006749

[1254 rows x 1 columns]

In [54]:
#Monthly returns with lamda aggregation: cumulative return calculation
fundsret_mtl = fundsret.resample('M').agg(lambda x: (x+1).prod() - 1)
fundsret_mtl

Ticker         FDGRX
Date                
2020-07-31  0.081224
2020-08-31  0.126135
2020-09-30 -0.033199
2020-10-31 -0.028562
2020-11-30  0.138751
2020-12-31  0.037423
2021-01-31  0.013995
2021-02-28  0.020702
2021-03-31 -0.010288
2021-04-30  0.064449
2021-05-31 -0.011719
2021-06-30  0.073687
2021-07-31  0.003681
2021-08-31  0.045062
2021-09-30 -0.050890
2021-10-31  0.082673
2021-11-30  0.017321
2021-12-31 -0.033866
2022-01-31 -0.111873
2022-02-28 -0.033344
2022-03-31  0.037368
2022-04-30 -0.144396
2022-05-31 -0.040662
2022-06-30 -0.086647
2022-07-31  0.119507
2022-08-31 -0.034850
2022-09-30 -0.099582
2022-10-31  0.065428
2022-11-30  0.049128
2022-12-31 -0.089017
2023-01-31  0.106175
2023-02-28 -0.012450
2023-03-31  0.067914
2023-04-30  0.009901
2023-05-31  0.071267
2023-06-30  0.064766
2023-07-31  0.041322
2023-08-31 -0.015238
2023-09-30 -0.064475
2023-10-31 -0.029979
2023-11-30  0.110835
2023-12-31  0.059996
2024-01-31  0.030388
2024-02-29  0.090909
2024-03-31  0.026756
2024-04-30 -0.044517
2024-05-31  0.074716
2024-06-30  0.062384
2024-07-31 -0.018910
2024-08-31  0.013695
2024-09-30  0.019014
2024-10-31  0.007611
2024-11-30  0.058479
2024-12-31  0.008032
2025-01-31  0.013634
2025-02-28 -0.043776
2025-03-31 -0.098210
2025-04-30  0.007941
2025-05-31  0.100169
2025-06-30  0.068286

In [23]:
%pip install pandas_datareader
import pandas_datareader.data as reader
import pandas as pd

Defaulting to user installation because normal site-packages is not writeable
Looking in links: /usr/share/pip-wheels
Note: you may need to restart the kernel to use updated packages.


In [49]:
factors = reader.DataReader('F-F_Research_Data_Factors', 'famafrench', start, end)[0]
factors

/home/8e6871ea-9e76-4029-849d-302a75dfa8e7/.local/lib/python3.11/site-packages/pandas_datareader/famafrench.py:114: FutureWarning: The argument 'date_parser' is deprecated and will be removed in a future version. Please use 'date_format' instead, or read your data in as 'object' dtype and then call 'to_datetime'.
  df = read_csv(StringIO("Date" + src[start:]), **params)
/home/8e6871ea-9e76-4029-849d-302a75dfa8e7/.local/lib/python3.11/site-packages/pandas_datareader/famafrench.py:114: FutureWarning: The argument 'date_parser' is deprecated and will be removed in a future version. Please use 'date_format' instead, or read your data in as 'object' dtype and then call 'to_datetime'.
  df = read_csv(StringIO("Date" + src[start:]), **params)


Mkt-RF   SMB    HML    RF
Date                              
2020-06    2.45  2.67  -2.17  0.01
2020-07    5.83 -2.35  -1.44  0.01
2020-08    7.62 -0.20  -3.02  0.01
2020-09   -3.64  0.06  -2.72  0.01
2020-10   -2.08  4.34   4.25  0.01
2020-11   12.45  5.81   2.09  0.01
2020-12    4.63  5.01  -1.68  0.01
2021-01   -0.07  7.14   3.22  0.00
2021-02    2.81  2.02   7.20  0.00
2021-03    3.16 -2.33   7.35  0.00
2021-04    4.95 -3.19  -1.02  0.00
2021-05    0.30 -0.20   7.04  0.00
2021-06    2.74  1.70  -7.86  0.00
2021-07    1.34 -3.93  -1.79  0.00
2021-08    2.94 -0.32  -0.23  0.00
2021-09   -4.40  0.66   5.12  0.00
2021-10    6.63 -2.27  -0.53  0.00
2021-11   -1.58 -1.31  -0.37  0.00
2021-12    3.23 -1.63   3.26  0.01
2022-01   -6.16 -5.93  12.86  0.00
2022-02   -2.28  2.16   3.06  0.00
2022-03    3.08 -1.55  -1.90  0.00
2022-04   -9.41 -1.35   6.18  0.00
2022-05   -0.34 -1.84   8.44  0.03
2022-06   -8.40  2.07  -6.00  0.06
2022-07    9.56  2.84  -4.15  0.08
2022-08   -3.77  1.41   0.18  0.19
2022-09   -9.34 -0.76   0.03  0.19
2022-10    7.85  0.01   7.97  0.23
2022-11    4.65 -3.57   1.48  0.29
2022-12   -6.37 -0.76   1.34  0.33
2023-01    6.61  5.01  -3.94  0.35
2023-02   -2.57  1.05  -0.80  0.34
2023-03    2.48 -5.48  -9.05  0.36
2023-04    0.62 -3.37  -0.12  0.35
2023-05    0.34  1.60  -7.79  0.36
2023-06    6.46  1.49  -0.13  0.40
2023-07    3.21  2.05   4.06  0.45
2023-08   -2.36 -3.14  -1.15  0.45
2023-09   -5.23 -2.51   1.49  0.43
2023-10   -3.15 -3.86   0.20  0.47
2023-11    8.88 -0.10   1.67  0.44
2023-12    4.85  6.30   4.89  0.43
2024-01    0.73 -5.07  -2.38  0.47
2024-02    5.07 -0.24  -3.43  0.42
2024-03    2.84 -2.50   4.16  0.43
2024-04   -4.65 -2.37  -0.45  0.47
2024-05    4.33  0.55  -1.31  0.44
2024-06    2.80 -3.00  -3.36  0.41
2024-07    1.22  6.81   5.62  0.45
2024-08    1.60 -3.49  -1.10  0.48
2024-09    1.72 -0.13  -2.77  0.40
2024-10   -1.00 -0.99   0.86  0.39
2024-11    6.49  4.46   0.15  0.40
2024-12   -3.17 -2.71  -3.00  0.37
2025-01    2.80 -1.97   1.63  0.37
2025-02   -2.44 -5.79   4.91  0.33
2025-03   -6.39 -2.76   2.90  0.34
2025-04   -0.84 -0.59  -3.40  0.35
2025-05    6.06  0.70  -2.88  0.38

In [51]:
#Deleting the first row
factors = factors[1:]
factors

Mkt-RF   SMB    HML    RF
Date                              
2020-07    5.83 -2.35  -1.44  0.01
2020-08    7.62 -0.20  -3.02  0.01
2020-09   -3.64  0.06  -2.72  0.01
2020-10   -2.08  4.34   4.25  0.01
2020-11   12.45  5.81   2.09  0.01
2020-12    4.63  5.01  -1.68  0.01
2021-01   -0.07  7.14   3.22  0.00
2021-02    2.81  2.02   7.20  0.00
2021-03    3.16 -2.33   7.35  0.00
2021-04    4.95 -3.19  -1.02  0.00
2021-05    0.30 -0.20   7.04  0.00
2021-06    2.74  1.70  -7.86  0.00
2021-07    1.34 -3.93  -1.79  0.00
2021-08    2.94 -0.32  -0.23  0.00
2021-09   -4.40  0.66   5.12  0.00
2021-10    6.63 -2.27  -0.53  0.00
2021-11   -1.58 -1.31  -0.37  0.00
2021-12    3.23 -1.63   3.26  0.01
2022-01   -6.16 -5.93  12.86  0.00
2022-02   -2.28  2.16   3.06  0.00
2022-03    3.08 -1.55  -1.90  0.00
2022-04   -9.41 -1.35   6.18  0.00
2022-05   -0.34 -1.84   8.44  0.03
2022-06   -8.40  2.07  -6.00  0.06
2022-07    9.56  2.84  -4.15  0.08
2022-08   -3.77  1.41   0.18  0.19
2022-09   -9.34 -0.76   0.03  0.19
2022-10    7.85  0.01   7.97  0.23
2022-11    4.65 -3.57   1.48  0.29
2022-12   -6.37 -0.76   1.34  0.33
2023-01    6.61  5.01  -3.94  0.35
2023-02   -2.57  1.05  -0.80  0.34
2023-03    2.48 -5.48  -9.05  0.36
2023-04    0.62 -3.37  -0.12  0.35
2023-05    0.34  1.60  -7.79  0.36
2023-06    6.46  1.49  -0.13  0.40
2023-07    3.21  2.05   4.06  0.45
2023-08   -2.36 -3.14  -1.15  0.45
2023-09   -5.23 -2.51   1.49  0.43
2023-10   -3.15 -3.86   0.20  0.47
2023-11    8.88 -0.10   1.67  0.44
2023-12    4.85  6.30   4.89  0.43
2024-01    0.73 -5.07  -2.38  0.47
2024-02    5.07 -0.24  -3.43  0.42
2024-03    2.84 -2.50   4.16  0.43
2024-04   -4.65 -2.37  -0.45  0.47
2024-05    4.33  0.55  -1.31  0.44
2024-06    2.80 -3.00  -3.36  0.41
2024-07    1.22  6.81   5.62  0.45
2024-08    1.60 -3.49  -1.10  0.48
2024-09    1.72 -0.13  -2.77  0.40
2024-10   -1.00 -0.99   0.86  0.39
2024-11    6.49  4.46   0.15  0.40
2024-12   -3.17 -2.71  -3.00  0.37
2025-01    2.80 -1.97   1.63  0.37
2025-02   -2.44 -5.79   4.91  0.33
2025-03   -6.39 -2.76   2.90  0.34
2025-04   -0.84 -0.59  -3.40  0.35
2025-05    6.06  0.70  -2.88  0.38

In [55]:
#Aligning the data set so that they both have the same length
fundsret_mtl = fundsret_mtl.iloc[:-1]
fundsret_mtl.shape

(59, 1)

In [52]:
factors.shape

(59, 4)

In [57]:
#because the dates are the same in both data frames although not the same format, we can still merge them, but first transform the data format for fundsmrl\_ret
fundsret_mtl.index = factors.index
fundsret_mtl

Ticker      FDGRX
Date             
2020-07  0.081224
2020-08  0.126135
2020-09 -0.033199
2020-10 -0.028562
2020-11  0.138751
2020-12  0.037423
2021-01  0.013995
2021-02  0.020702
2021-03 -0.010288
2021-04  0.064449
2021-05 -0.011719
2021-06  0.073687
2021-07  0.003681
2021-08  0.045062
2021-09 -0.050890
2021-10  0.082673
2021-11  0.017321
2021-12 -0.033866
2022-01 -0.111873
2022-02 -0.033344
2022-03  0.037368
2022-04 -0.144396
2022-05 -0.040662
2022-06 -0.086647
2022-07  0.119507
2022-08 -0.034850
2022-09 -0.099582
2022-10  0.065428
2022-11  0.049128
2022-12 -0.089017
2023-01  0.106175
2023-02 -0.012450
2023-03  0.067914
2023-04  0.009901
2023-05  0.071267
2023-06  0.064766
2023-07  0.041322
2023-08 -0.015238
2023-09 -0.064475
2023-10 -0.029979
2023-11  0.110835
2023-12  0.059996
2024-01  0.030388
2024-02  0.090909
2024-03  0.026756
2024-04 -0.044517
2024-05  0.074716
2024-06  0.062384
2024-07 -0.018910
2024-08  0.013695
2024-09  0.019014
2024-10  0.007611
2024-11  0.058479
2024-12  0.008032
2025-01  0.013634
2025-02 -0.043776
2025-03 -0.098210
2025-04  0.007941
2025-05  0.100169

In [68]:
merge = pd.merge(fundsret_mtl, factors, on = 'Date')
merge
#important to know that FDGRX is decimal, everything else is in %

FDGRX  Mkt-RF   SMB    HML    RF
Date                                        
2020-07  0.081224    5.83 -2.35  -1.44  0.01
2020-08  0.126135    7.62 -0.20  -3.02  0.01
2020-09 -0.033199   -3.64  0.06  -2.72  0.01
2020-10 -0.028562   -2.08  4.34   4.25  0.01
2020-11  0.138751   12.45  5.81   2.09  0.01
2020-12  0.037423    4.63  5.01  -1.68  0.01
2021-01  0.013995   -0.07  7.14   3.22  0.00
2021-02  0.020702    2.81  2.02   7.20  0.00
2021-03 -0.010288    3.16 -2.33   7.35  0.00
2021-04  0.064449    4.95 -3.19  -1.02  0.00
2021-05 -0.011719    0.30 -0.20   7.04  0.00
2021-06  0.073687    2.74  1.70  -7.86  0.00
2021-07  0.003681    1.34 -3.93  -1.79  0.00
2021-08  0.045062    2.94 -0.32  -0.23  0.00
2021-09 -0.050890   -4.40  0.66   5.12  0.00
2021-10  0.082673    6.63 -2.27  -0.53  0.00
2021-11  0.017321   -1.58 -1.31  -0.37  0.00
2021-12 -0.033866    3.23 -1.63   3.26  0.01
2022-01 -0.111873   -6.16 -5.93  12.86  0.00
2022-02 -0.033344   -2.28  2.16   3.06  0.00
2022-03  0.037368    3.08 -1.55  -1.90  0.00
2022-04 -0.144396   -9.41 -1.35   6.18  0.00
2022-05 -0.040662   -0.34 -1.84   8.44  0.03
2022-06 -0.086647   -8.40  2.07  -6.00  0.06
2022-07  0.119507    9.56  2.84  -4.15  0.08
2022-08 -0.034850   -3.77  1.41   0.18  0.19
2022-09 -0.099582   -9.34 -0.76   0.03  0.19
2022-10  0.065428    7.85  0.01   7.97  0.23
2022-11  0.049128    4.65 -3.57   1.48  0.29
2022-12 -0.089017   -6.37 -0.76   1.34  0.33
2023-01  0.106175    6.61  5.01  -3.94  0.35
2023-02 -0.012450   -2.57  1.05  -0.80  0.34
2023-03  0.067914    2.48 -5.48  -9.05  0.36
2023-04  0.009901    0.62 -3.37  -0.12  0.35
2023-05  0.071267    0.34  1.60  -7.79  0.36
2023-06  0.064766    6.46  1.49  -0.13  0.40
2023-07  0.041322    3.21  2.05   4.06  0.45
2023-08 -0.015238   -2.36 -3.14  -1.15  0.45
2023-09 -0.064475   -5.23 -2.51   1.49  0.43
2023-10 -0.029979   -3.15 -3.86   0.20  0.47
2023-11  0.110835    8.88 -0.10   1.67  0.44
2023-12  0.059996    4.85  6.30   4.89  0.43
2024-01  0.030388    0.73 -5.07  -2.38  0.47
2024-02  0.090909    5.07 -0.24  -3.43  0.42
2024-03  0.026756    2.84 -2.50   4.16  0.43
2024-04 -0.044517   -4.65 -2.37  -0.45  0.47
2024-05  0.074716    4.33  0.55  -1.31  0.44
2024-06  0.062384    2.80 -3.00  -3.36  0.41
2024-07 -0.018910    1.22  6.81   5.62  0.45
2024-08  0.013695    1.60 -3.49  -1.10  0.48
2024-09  0.019014    1.72 -0.13  -2.77  0.40
2024-10  0.007611   -1.00 -0.99   0.86  0.39
2024-11  0.058479    6.49  4.46   0.15  0.40
2024-12  0.008032   -3.17 -2.71  -3.00  0.37
2025-01  0.013634    2.80 -1.97   1.63  0.37
2025-02 -0.043776   -2.44 -5.79   4.91  0.33
2025-03 -0.098210   -6.39 -2.76   2.90  0.34
2025-04  0.007941   -0.84 -0.59  -3.40  0.35
2025-05  0.100169    6.06  0.70  -2.88  0.38

In [76]:
#resigning and dividing 
merge[['Mkt-RF', 'SMB', 'HML', 'RF']] = merge[['Mkt-RF', 'SMB', 'HML', 'RF']]/100
merge

FDGRX  Mkt-RF     SMB     HML      RF  FDGRX-RF
Date                                                       
2020-07  0.081224  0.0583 -0.0235 -0.0144  0.0001  0.071224
2020-08  0.126135  0.0762 -0.0020 -0.0302  0.0001  0.116135
2020-09 -0.033199 -0.0364  0.0006 -0.0272  0.0001 -0.043199
2020-10 -0.028562 -0.0208  0.0434  0.0425  0.0001 -0.038562
2020-11  0.138751  0.1245  0.0581  0.0209  0.0001  0.128751
2020-12  0.037423  0.0463  0.0501 -0.0168  0.0001  0.027423
2021-01  0.013995 -0.0007  0.0714  0.0322  0.0000  0.013995
2021-02  0.020702  0.0281  0.0202  0.0720  0.0000  0.020702
2021-03 -0.010288  0.0316 -0.0233  0.0735  0.0000 -0.010288
2021-04  0.064449  0.0495 -0.0319 -0.0102  0.0000  0.064449
2021-05 -0.011719  0.0030 -0.0020  0.0704  0.0000 -0.011719
2021-06  0.073687  0.0274  0.0170 -0.0786  0.0000  0.073687
2021-07  0.003681  0.0134 -0.0393 -0.0179  0.0000  0.003681
2021-08  0.045062  0.0294 -0.0032 -0.0023  0.0000  0.045062
2021-09 -0.050890 -0.0440  0.0066  0.0512  0.0000 -0.050890
2021-10  0.082673  0.0663 -0.0227 -0.0053  0.0000  0.082673
2021-11  0.017321 -0.0158 -0.0131 -0.0037  0.0000  0.017321
2021-12 -0.033866  0.0323 -0.0163  0.0326  0.0001 -0.043866
2022-01 -0.111873 -0.0616 -0.0593  0.1286  0.0000 -0.111873
2022-02 -0.033344 -0.0228  0.0216  0.0306  0.0000 -0.033344
2022-03  0.037368  0.0308 -0.0155 -0.0190  0.0000  0.037368
2022-04 -0.144396 -0.0941 -0.0135  0.0618  0.0000 -0.144396
2022-05 -0.040662 -0.0034 -0.0184  0.0844  0.0003 -0.070662
2022-06 -0.086647 -0.0840  0.0207 -0.0600  0.0006 -0.146647
2022-07  0.119507  0.0956  0.0284 -0.0415  0.0008  0.039507
2022-08 -0.034850 -0.0377  0.0141  0.0018  0.0019 -0.224850
2022-09 -0.099582 -0.0934 -0.0076  0.0003  0.0019 -0.289582
2022-10  0.065428  0.0785  0.0001  0.0797  0.0023 -0.164572
2022-11  0.049128  0.0465 -0.0357  0.0148  0.0029 -0.240872
2022-12 -0.089017 -0.0637 -0.0076  0.0134  0.0033 -0.419017
2023-01  0.106175  0.0661  0.0501 -0.0394  0.0035 -0.243825
2023-02 -0.012450 -0.0257  0.0105 -0.0080  0.0034 -0.352450
2023-03  0.067914  0.0248 -0.0548 -0.0905  0.0036 -0.292086
2023-04  0.009901  0.0062 -0.0337 -0.0012  0.0035 -0.340099
2023-05  0.071267  0.0034  0.0160 -0.0779  0.0036 -0.288733
2023-06  0.064766  0.0646  0.0149 -0.0013  0.0040 -0.335234
2023-07  0.041322  0.0321  0.0205  0.0406  0.0045 -0.408678
2023-08 -0.015238 -0.0236 -0.0314 -0.0115  0.0045 -0.465238
2023-09 -0.064475 -0.0523 -0.0251  0.0149  0.0043 -0.494475
2023-10 -0.029979 -0.0315 -0.0386  0.0020  0.0047 -0.499979
2023-11  0.110835  0.0888 -0.0010  0.0167  0.0044 -0.329165
2023-12  0.059996  0.0485  0.0630  0.0489  0.0043 -0.370004
2024-01  0.030388  0.0073 -0.0507 -0.0238  0.0047 -0.439612
2024-02  0.090909  0.0507 -0.0024 -0.0343  0.0042 -0.329091
2024-03  0.026756  0.0284 -0.0250  0.0416  0.0043 -0.403244
2024-04 -0.044517 -0.0465 -0.0237 -0.0045  0.0047 -0.514517
2024-05  0.074716  0.0433  0.0055 -0.0131  0.0044 -0.365284
2024-06  0.062384  0.0280 -0.0300 -0.0336  0.0041 -0.347616
2024-07 -0.018910  0.0122  0.0681  0.0562  0.0045 -0.468910
2024-08  0.013695  0.0160 -0.0349 -0.0110  0.0048 -0.466305
2024-09  0.019014  0.0172 -0.0013 -0.0277  0.0040 -0.380986
2024-10  0.007611 -0.0100 -0.0099  0.0086  0.0039 -0.382389
2024-11  0.058479  0.0649  0.0446  0.0015  0.0040 -0.341521
2024-12  0.008032 -0.0317 -0.0271 -0.0300  0.0037 -0.361968
2025-01  0.013634  0.0280 -0.0197  0.0163  0.0037 -0.356366
2025-02 -0.043776 -0.0244 -0.0579  0.0491  0.0033 -0.373776
2025-03 -0.098210 -0.0639 -0.0276  0.0290  0.0034 -0.438210
2025-04  0.007941 -0.0084 -0.0059 -0.0340  0.0035 -0.342059
2025-05  0.100169  0.0606  0.0070 -0.0288  0.0038 -0.279831

In [77]:
# creating the dependent variable for the regression r_i - r_f
merge['FDGRX-RF'] = merge.FDGRX - merge.RF

In [78]:
merge

FDGRX  Mkt-RF     SMB     HML      RF  FDGRX-RF
Date                                                       
2020-07  0.081224  0.0583 -0.0235 -0.0144  0.0001  0.081124
2020-08  0.126135  0.0762 -0.0020 -0.0302  0.0001  0.126035
2020-09 -0.033199 -0.0364  0.0006 -0.0272  0.0001 -0.033299
2020-10 -0.028562 -0.0208  0.0434  0.0425  0.0001 -0.028662
2020-11  0.138751  0.1245  0.0581  0.0209  0.0001  0.138651
2020-12  0.037423  0.0463  0.0501 -0.0168  0.0001  0.037323
2021-01  0.013995 -0.0007  0.0714  0.0322  0.0000  0.013995
2021-02  0.020702  0.0281  0.0202  0.0720  0.0000  0.020702
2021-03 -0.010288  0.0316 -0.0233  0.0735  0.0000 -0.010288
2021-04  0.064449  0.0495 -0.0319 -0.0102  0.0000  0.064449
2021-05 -0.011719  0.0030 -0.0020  0.0704  0.0000 -0.011719
2021-06  0.073687  0.0274  0.0170 -0.0786  0.0000  0.073687
2021-07  0.003681  0.0134 -0.0393 -0.0179  0.0000  0.003681
2021-08  0.045062  0.0294 -0.0032 -0.0023  0.0000  0.045062
2021-09 -0.050890 -0.0440  0.0066  0.0512  0.0000 -0.050890
2021-10  0.082673  0.0663 -0.0227 -0.0053  0.0000  0.082673
2021-11  0.017321 -0.0158 -0.0131 -0.0037  0.0000  0.017321
2021-12 -0.033866  0.0323 -0.0163  0.0326  0.0001 -0.033966
2022-01 -0.111873 -0.0616 -0.0593  0.1286  0.0000 -0.111873
2022-02 -0.033344 -0.0228  0.0216  0.0306  0.0000 -0.033344
2022-03  0.037368  0.0308 -0.0155 -0.0190  0.0000  0.037368
2022-04 -0.144396 -0.0941 -0.0135  0.0618  0.0000 -0.144396
2022-05 -0.040662 -0.0034 -0.0184  0.0844  0.0003 -0.040962
2022-06 -0.086647 -0.0840  0.0207 -0.0600  0.0006 -0.087247
2022-07  0.119507  0.0956  0.0284 -0.0415  0.0008  0.118707
2022-08 -0.034850 -0.0377  0.0141  0.0018  0.0019 -0.036750
2022-09 -0.099582 -0.0934 -0.0076  0.0003  0.0019 -0.101482
2022-10  0.065428  0.0785  0.0001  0.0797  0.0023  0.063128
2022-11  0.049128  0.0465 -0.0357  0.0148  0.0029  0.046228
2022-12 -0.089017 -0.0637 -0.0076  0.0134  0.0033 -0.092317
2023-01  0.106175  0.0661  0.0501 -0.0394  0.0035  0.102675
2023-02 -0.012450 -0.0257  0.0105 -0.0080  0.0034 -0.015850
2023-03  0.067914  0.0248 -0.0548 -0.0905  0.0036  0.064314
2023-04  0.009901  0.0062 -0.0337 -0.0012  0.0035  0.006401
2023-05  0.071267  0.0034  0.0160 -0.0779  0.0036  0.067667
2023-06  0.064766  0.0646  0.0149 -0.0013  0.0040  0.060766
2023-07  0.041322  0.0321  0.0205  0.0406  0.0045  0.036822
2023-08 -0.015238 -0.0236 -0.0314 -0.0115  0.0045 -0.019738
2023-09 -0.064475 -0.0523 -0.0251  0.0149  0.0043 -0.068775
2023-10 -0.029979 -0.0315 -0.0386  0.0020  0.0047 -0.034679
2023-11  0.110835  0.0888 -0.0010  0.0167  0.0044  0.106435
2023-12  0.059996  0.0485  0.0630  0.0489  0.0043  0.055696
2024-01  0.030388  0.0073 -0.0507 -0.0238  0.0047  0.025688
2024-02  0.090909  0.0507 -0.0024 -0.0343  0.0042  0.086709
2024-03  0.026756  0.0284 -0.0250  0.0416  0.0043  0.022456
2024-04 -0.044517 -0.0465 -0.0237 -0.0045  0.0047 -0.049217
2024-05  0.074716  0.0433  0.0055 -0.0131  0.0044  0.070316
2024-06  0.062384  0.0280 -0.0300 -0.0336  0.0041  0.058284
2024-07 -0.018910  0.0122  0.0681  0.0562  0.0045 -0.023410
2024-08  0.013695  0.0160 -0.0349 -0.0110  0.0048  0.008895
2024-09  0.019014  0.0172 -0.0013 -0.0277  0.0040  0.015014
2024-10  0.007611 -0.0100 -0.0099  0.0086  0.0039  0.003711
2024-11  0.058479  0.0649  0.0446  0.0015  0.0040  0.054479
2024-12  0.008032 -0.0317 -0.0271 -0.0300  0.0037  0.004332
2025-01  0.013634  0.0280 -0.0197  0.0163  0.0037  0.009934
2025-02 -0.043776 -0.0244 -0.0579  0.0491  0.0033 -0.047076
2025-03 -0.098210 -0.0639 -0.0276  0.0290  0.0034 -0.101610
2025-04  0.007941 -0.0084 -0.0059 -0.0340  0.0035  0.004441
2025-05  0.100169  0.0606  0.0070 -0.0288  0.0038  0.096369

In [79]:
y = merge['FDGRX-RF']
X = merge[['Mkt-RF', 'SMB', 'HML']]
#adding the const
X_sm = sm.add_constant(X)

In [80]:
model = sm.OLS(y, X_sm)
results = model.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               FDGRX-RF   R-squared:                       0.938
Model:                            OLS   Adj. R-squared:                  0.934
Method:                 Least Squares   F-statistic:                     276.1
Date:                Thu, 24 Jul 2025   Prob (F-statistic):           4.06e-33
Time:                        07:49:11   Log-Likelihood:                 161.20
No. Observations:                  59   AIC:                            -314.4
Df Residuals:                      55   BIC:                            -306.1
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0032      0.002      1.438      0.156      -0.001       0.008
Mkt-RF         1.1409      0.046     24.551      0.000       1.048       1.234
SMB            0.0628      0.071      0.888      0.378      -0.079       0.204
HML           -0.4310      0.051     -8.491      0.000      -0.533      -0.329
==============================================================================
Omnibus:                       12.309   Durbin-Watson:                   2.596
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               14.581
Skew:                          -0.844   Prob(JB):                     0.000682
Kurtosis:                       4.755   Cond. No.                         34.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

# R squared is nice, almost 1
## Market risk ~ beta, 1.14 , fidelity fund has more market risk than the AVG asset
## Size factor slightly more positive, tend to invest rather in smaller stocks 
## Large negative loading, fidelity growth fund overweighting growth stocks relative to the whole market. 